### PR TITLE
updated interface to define has_edge, adapted SimpleGraphs

### PR DESCRIPTION
**Pull request by matbesancon**
following the discussion #939, this is an update to the interface and documentation.
The `has_edge` methods are also re-defined for SimpleGraphs

### CODE REVIEW

File: src/SimpleGraphs/simpledigraph.jl [hunk 1]
```diff
@@ -245,9 +245,7 @@ is_directed(g::SimpleDiGraph) = true
 is_directed(::Type{SimpleDiGraph}) = true
 is_directed(::Type{SimpleDiGraph{T}}) where T = true
 
-
-function has_edge(g::SimpleDiGraph{T}, e::SimpleDiGraphEdge{T}) where T
-    s, d = T.(Tuple(e))
+function has_edge(g::SimpleDiGraph{T}, s, d) where T
     verts = vertices(g)
     (s in verts && d in verts) || return false  # edge out of bounds
     @inbounds list = g.fadjlist[s]
```
Comment (sbromberger) [inline src/SimpleGraphs/simpledigraph.jl]: why is this `@inbounds`? I don't think it matters much. Do benchmarks say otherwise?

(I see we're doing it below, but we should avoid annotations if they don't do anything.)

File: src/interface.jl [hunk 2]
```diff
@@ -130,17 +130,21 @@ is_directed(::Type{T}) where T = _NI("is_directed")
 
 Return true if `v` is a vertex of `g`.
 """
-has_vertex(x, v) = _NI("has_vertex")
+has_vertex(g, v) = _NI("has_vertex")
```
Comment (sbromberger) [inline src/interface.jl]: Let's revert this. As discussed, `x` is used deliberately since it does not restrict itself to graphs. This function gets called on any two arguments.

File: src/interface.jl [hunk 3]
```diff
@@ -150,7 +154,7 @@ Return a list of all neighbors connected to vertex `v` by an incoming edge.
 ### Implementation Notes
 Returns a reference, not a copy. Do not modify result.
 """
-inneighbors(x, v) = _NI("inneighbors")
+inneighbors(g, v) = _NI("inneighbors")
```
Comment (sbromberger) [inline src/interface.jl]: see above. Also, this is unrelated to the PR.

File: src/interface.jl [hunk 4]
```diff
@@ -160,7 +164,7 @@ Return a list of all neighbors connected to vertex `v` by an outgoing edge.
 # Implementation Notes
 Returns a reference, not a copy. Do not modify result.
 """
-outneighbors(x, v) = _NI("outneighbors")
+outneighbors(g, v) = _NI("outneighbors")
```
Comment (sbromberger) [inline src/interface.jl]: as above

File: docs/src/basicproperties.md [hunk 5]
```diff
@@ -8,7 +8,7 @@ The following is an overview of functions for accessing graph properties. For fu
 - `vertices`: Iterable object of all graph vertices.
 - `edges`: Iterable object of all graph edges.
 - `has_vertex`: Checks for whether graph includes a vertex.
-- `has_edge`: Checks for whether graph includes an edge.
+- `has_edge`: Checks for whether graph includes an edge from a given source to a given desination.
```
Comment (sbromberger) [inline docs/src/basicproperties.md]: (Note typo in `destination`.) Would suggest: 

```
Checks for whether graph includes an edge from a given source to a given destination.
In most cases, `has_edge` can also check whether a graph includes an edge `e` that is
equal to another edge in the graph (equality being defined by the specific graph type).
```

Comment (matbesancon) [reply docs/src/basicproperties.md]: why "in most cases", if you don't define it, it has a default implementation so it will check it

Comment (matbesancon) [reply docs/src/basicproperties.md]: so "`has_edge` can also check..." works fine

Comment (sbromberger) [reply docs/src/basicproperties.md]: I think we need more explanatory text - either here, or somewhere else, along these lines:

> `has_edge(g, i, j)` will return `true` if there is an edge between vertices `i` and `j` in graph `g`. This version of `has_edge` ignores any other metadata (such as weight) associated with the edge.
> `has_edge(g, e)` will return `true` if there is an edge in `g` that satisfies `e == f` for any `f ∈ edges(g)`. This is a strict equality test that may not limit itself for testing that just there exists any edge between the source and destination.
> Note: to use the `has_edge(g, e)` method safely, it is important to understand the conditions under which edges are equal to each other. This will necessarily vary among edge types and defaults to `src(e) == src(f) && dst(e) == dst(f)`.

File: docs/src/basicproperties.md [hunk 6]
```diff
@@ -8,7 +8,9 @@ The following is an overview of functions for accessing graph properties. For fu
 - `vertices`: Iterable object of all graph vertices.
 - `edges`: Iterable object of all graph edges.
 - `has_vertex`: Checks for whether graph includes a vertex.
-- `has_edge`: Checks for whether graph includes an edge.
+- `has_edge(g, s, d)`: Checks for whether graph includes an edge from a given source `s` to a given destination `d`.
+-  `has_edge(g, e)` will return true if there is an edge in g that satisfies `e == f` for any `f ∈ edges(g)`. This is a strict equality test that may not limit itself for testing that just there exists any edge between the source and destination.
```
Comment (sbromberger) [inline docs/src/basicproperties.md]: I think I understand what you're trying to say here, but it's a bit unclear. Perhaps
> This is a strict equality test that, depending on the implementation of the edge type, may be broader than checking that an edge exists between `src(e)` and `dst(e)`.

Comment (matbesancon) [reply docs/src/basicproperties.md]: sorry just copied your previous text

Comment (sbromberger) [reply docs/src/basicproperties.md]: Yeah. The fact that I didn't recognize the text means that it's bad :) Sorry about that.
